Qualcomm AI Engine Direct - Quantizer refine for qat

chunit-quic opened the pull request:
- Reorginize qualcomm/quantizer
- Split quantizer/utils.py to 
-- qconfig
-- annotators
-- observers directory
- Change coresponding callees
- Rename get_default_Nbit_qnn_ptq_config to get_NaNw_qnn_ptq_config
- Add 16a4w conv test* (It is not compared with original model)

Commit:
[Qualcomm AI Engine Direct - Quantizer refine for qat]

- Reorginize qualcomm/quantizer
- Split quantizer/utils.py to
-- qconfig
-- annotators
-- observers directory
- Change coresponding callees
- Rename get_default_Nbit_qnn_ptq_config to get_NaNw_qnn_ptq_config
- Add 16a4w conv test* (It is not compared with original model)

chunit-quic commented:
Hi @cccclai,

We implement the draft PR for
1. Refine API of QnnQauntizer
2. Add more qat quant confings
3. Add 16a4w qat conv test case

We have another local branch which is testing QAT mobilenet v2 based on it. Yet we encounter some problems. 
During QAT process, the the accuracy/loss seems to be reasonable if we only quantize conv ops. Yet if we quantize any other op, accuracy/loss drop significantly. Is there any known issue for this? Or maybe it just results from some missetting form us. (We can make a patch or update the local branch here if you are insterested in it)

 Thank you. :)

cccclai commented:
@navsud can you review this PR?

navsud commented on backends/qualcomm/quantizer/observers/param_observer.py:
```diff
@@ -0,0 +1,103 @@
+from torch.ao.quantization.observer import UniformQuantizationObserverBase
+
+
+class ParamObserver(UniformQuantizationObserverBase):
```
This looks great and is generic enough to be added into torch/ao/quantization/observer.py. If possible, please move it there as a follow-up. 

How about renaming it to PerChannelParamObserver() or PerChannelWeightObserver?

chunit-quic replied:
It's great to see you like it! Change the name and add a TODO for it

navsud commented on backends/qualcomm/quantizer/qconfig.py:
```diff
@@ -0,0 +1,463 @@
+        qscheme=torch.per_tensor_symmetric,
+        ch_axis=0,
+        reduce_range=True,
+        observer=MovingAverageMinMaxObserver,
```
This should be PerChannelMovingAverageMinMaxObserver and qscheme=torch.per_channel_symmetric?

chunit-quic replied:
Sorry for misunderstanding. This quant config is for per tensor specifically. The per channel one is [here](https://github.com/pytorch/executorch/pull/6513/files#diff-c990fa4562a79a98b35cc2429e09dc2461e175cc7bf0a8b112655fef021dc5edR400), and we use[ the condition of a quantizer member function here](https://github.com/pytorch/executorch/pull/6513/files#diff-4cb9d05b0ef6711a9bc59c1556fd35ee441ece67389052a740117f9d1cd04a0cR122) to assign per channel quant confing.

navsud commented on backends/qualcomm/quantizer/quantizer.py:
```diff
@@ -22,25 +22,29 @@
-    get_default_8bit_qat_proto,
-    get_default_8bit_qnn_ptq_config,
+    get_8a8w_qnn_ptq_config,
+    get_8a8w_qnn_qat_config,
```
As a follow-up PR, can we unify/simplify: `get_8a8w_qnn_ptq_config()` and `get_8a8w_qnn_qat_config()` into `get_8a8w_qnn_config()` with the argument is_train=True/False, which will define whether we are doing PTQ vs. QAT.

chunit-quic replied:
No problem. We would like to keep separated at first. Once it seems to be stable we will raise a TODO PR.

navsud reviewed (commented):
Overall LGTM.
cc: @cccclai for final approval.

Commit:
Fix baed on comments

- Move and rename param_observer.py to per_channel_param_observer.py
- Add todo to merge qconfig

Commit:
Add a comment

- Add todo for per_channel_param_observer.py

chunit-quic commented:
> Overall LGTM. cc: @cccclai for final approval.

Hi @navsud

Thank you very much for reviewing! We just uploaded a patch based on comments. Just two questions we would like to ask.

1. Does QAT only work for ops with weight at this moment? I tried to do QAT for a single op add, and got complained about no parameters to be trained. Is it expected?
2. Similar to the point 1. We are not pretty sure whether we quantize ops of mobilenetV2 properly. Accuracy drop significantly during QAT if we add quant config to any op but convolution op. More details can be found [in the first comment of the PR](https://github.com/pytorch/executorch/pull/6513#issuecomment-2440510907)

Thank you for reading! If anything is unclear please feel free to let me know. I will try to describe it more. :D

Commit:
[Fix lint]

navsud reviewed (approved):
LGTM

chunit-quic commented:
Hi @cccclai, may you take a glance at this PR when you are free, and perhas merge it, if it looks good to you? Thanks :D

cccclai reviewed (approved):
Looks good to me - thank you for adding the feature!

cccclai commented:
I notice that I didn't import to internal...in the worst case, if this diff breaks some tests, we may need to revert and reland it...

chunit-quic commented:
> I notice that I didn't import to internal...in the worst case, if this diff breaks some tests, we may need to revert and reland it...

It's fine. If so just let me know what should I fix and I will do it asap.  :)

cccclai commented:
@chunit-quic Unfortunately it's reverted because it breaks internal test....can you submit a PR again?

chunit-quic commented:
> @chunit-quic Unfortunately it's reverted because it breaks internal test....can you submit a PR again?

No problem. The new PR is [PR6747](https://github.com/pytorch/executorch/pull/6747)
